rusb-async/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+use rusb::ffi::{self, constants::*};
+
+use std::convert::TryInto;
+use std::ptr::NonNull;
+
+use std::sync::atomic::{AtomicBool, Ordering};
+
+mod error;
+mod pool;
+
+use error::{Error, Result};
+
+pub use pool::TransferPool;
+
+struct Transfer {
+    ptr: NonNull<ffi::libusb_transfer>,
+    buffer: Vec<u8>,
+}
+
+impl Transfer {
+    // Invariant: Caller must ensure `device` outlives this transfer
+    unsafe fn bulk(
+        device: *mut ffi::libusb_device_handle,
+        endpoint: u8,
+        mut buffer: Vec<u8>,
+    ) -> Self {
+        // non-isochronous endpoints (e.g. control, bulk, interrupt) specify a value of 0
+        // This is step 1 of async API
+
+        let ptr =
+            NonNull::new(ffi::libusb_alloc_transfer(0)).expect("Could not allocate transfer!");
+
+        let user_data = Box::into_raw(Box::new(AtomicBool::new(false))).cast::<libc::c_void>();
+
+        let length = if endpoint & ffi::constants::LIBUSB_ENDPOINT_DIR_MASK
+            == ffi::constants::LIBUSB_ENDPOINT_OUT
+        {
+            // for OUT endpoints: the currently valid data in the buffer
+            buffer.len()
+        } else {
+            // for IN endpoints: the full capacity
+            buffer.capacity()
+        }
+        .try_into()
+        .unwrap();
+
+        ffi::libusb_fill_bulk_transfer(
+            ptr.as_ptr(),
+            device,
+            endpoint,
+            buffer.as_mut_ptr(),
+            length,
+            Self::transfer_cb,
+            user_data,
+            0,
+        );
+
+        Self { ptr, buffer }
+    }
+
+    // Invariant: Caller must ensure `device` outlives this transfer
+    unsafe fn control(
+        device: *mut ffi::libusb_device_handle,
+
+        request_type: u8,
+        request: u8,
+        value: u16,
+        index: u16,
+        data: &[u8],
+    ) -> Self {
+        let mut buf = Vec::with_capacity(data.len() + LIBUSB_CONTROL_SETUP_SIZE);
```

Comment:
Sooo... how is data actually used? You create the buffer with it's length, and you pass it's length to `control_setup`, but where do you actually copy it into the buffer?

Reply:
I'm working on the other part of my refactor for iso transfers right now, but someone on the Rust discord suggested boxed slices instead of this weird Vec stuff.
BTW I don't think the length is ever updated either, so all the data in the buffer is in the reserved portion (as it's only written to by the FFI function which doesn't know about Vecs